### PR TITLE
Add Quarkus Vault Authentication Guide

### LINKED ISSUE #4846: Quarkus Vault Authentication Guide
**Description**
Provide a guide showing how to deploy hashicorp vault and a client quarkus application in kubernetes, demonstrating the [vault kubernetes auth method](https://www.vaultproject.io/docs/auth/kubernetes.html).

**Implementation ideas**
see the [vault quarkus kubernetes tutorial](https://github.com/vsevel/vaultapp)

For a more comprehensive guide, all vault authentication methods currently supported in Quarkus will be documented.

### PR DESCRIPTION
**Pull request by vsevel**
This provides a dedicated guide that covers all currently supported Vault Authentication methods:
 - direct token (including variant with response wrapping)
 - userpass
 - approle (including variant with response wrapping)
 - kubernetes

Fixes #4846

### CODE REVIEW

**Review by lordofthejars — APPROVED** [review]
Really nice document that will allow us to add more strategies in the future.

**Comment by lordofthejars** [comment]
@vsevel one comment shouldn't this PR contains removing the lines of the previous docs?

**Comment by sberyozkin** [comment]
@lordofthejars 

> one comment shouldn't this PR contains removing the lines of the previous docs?

Can you clarify please ? Thanks

**Comment by lordofthejars** [comment]
@sberyozkin I mean that some of the content added in this doc is duplicated in other Vault guides, so maybe they should be removed as well, and this PR could be used for this.

**Comment by sberyozkin** [comment]
@lordofthejars Good point, I suppose the those docs can provide some basic relevant info with the link to the main Vault authentication doc

**Comment by vsevel** [comment]
@lordofthejars @sberyozkin thanks for your encouraging comments :-)

> Really nice document that will allow us to add more strategies in the future.

yes definitely. for instance for the cert auth.

> I mean that some of the content added in this doc is duplicated in other Vault guides, so 
maybe they should be removed as well, and this PR could be used for this.

actually there is almost no overlap, because in the base guide I activate the `userpass` auth method and set up the user with a note specifying that other auth methods exist but without any details (client token, approle and k8s); and in the vault-auth guide I say in the `userpass` section that I expect the `userpass` method to be already enabled from the base guide. if we consider the quarkus documentation to be guides (and not reference doc), then I think it is fine.

also in the vault-auth guide I set the base guide to be a prerequisite (allowing to reuse the configured vault + the app). because I think the base guide is what people will want to discover first. and we need to provide a minimal vault auth setup (`userpass` in this case). the base guide needs to be standalone.

one thing we could do in the base guide is to remove the _The Vault extension also supports alternate authentication methods such as:_ note and replace it with: _Other auth methods are supported. Check the vault auth guide for details_.

what yo you think?

File: docs/src/main/asciidoc/vault-auth.adoc [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+:client-token: s.JqMeE1UEyUb19F6zmMW0SWx6
+:client-token-wrapping-token: s.2cLMBoKhelDK6W3uAFT2umXu
+:role-id: b15460ff-fea0-43fc-1002-a045fb60dfc4
+:secret-id: d2f13e1f-f32a-f60a-86d8-0b5cdaeb821b
+:secret-id-wrapping-token: s.aSq7tcRqfeboZqLMPfa5gkXJ
+
+Working with Vault is typically a 2 step process:
+
+* Logging in, which returns a client token
+* Start using Vault using the client token, within the limits of what is allowed in policies associated with the token
```
Comment (gsmet) [inline docs/src/main/asciidoc/vault-auth.adoc:23]:
```suggestion
* Start using Vault using the client token, within the limits of what is allowed by the policies associated with the token
```
?

File: docs/src/main/asciidoc/vault-auth.adoc [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+
+To complete this guide, you need:
+
+* to complete the {base-guide}
+* roughly 30 minutes
+* an IDE
+* JDK 1.8+ installed with `JAVA_HOME` configured appropriately
+* Apache Maven {maven-version}
+* Docker installed
+* For the Kubernetes authentication: A Kubernetes distribution to deploy the Quarkus application (minishift, k3s, docker desktop, …​)
```
Comment (gsmet) [inline docs/src/main/asciidoc/vault-auth.adoc:47]:
```suggestion
* For the Kubernetes authentication: A Kubernetes distribution to deploy the Quarkus application (Minishift, K3s, Docker Desktop, …​)
```

File: docs/src/main/asciidoc/vault-auth.adoc [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+
+== Token Authentication
+
+We assume there is a Vault container and a PostgresQL container running from the {base-guide}, and the root token is known.
```
Comment (gsmet) [inline docs/src/main/asciidoc/vault-auth.adoc:51]:
```suggestion
We assume there is a Vault container and a PostgreSQL container running from the {base-guide}, and the root token is known.
```

File: docs/src/main/asciidoc/vault-auth.adoc [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+
+First, create a new shell, `docker exec` into the container and set the root token:
+[source, shell, subs=attributes+]
+----
+docker exec -it dev-vault sh
+/ # export VAULT_TOKEN={root-token}
+----
+
+Create a token for policy `vault-quickstart-policy`:
```
Comment (gsmet) [inline docs/src/main/asciidoc/vault-auth.adoc:60]:
```suggestion
Create a token for the `vault-quickstart-policy` policy:
```

File: docs/src/main/asciidoc/vault-auth.adoc [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+[source, shell, subs=attributes+]
+----
+curl http://localhost:8080/hello/private-key
+----
+
+You should see: `{private-key}`.
+
+=== Client Token using Response Wrapping
+
+One drawback of this approach is that you expose a secret piece of information (.i.e the token) that can give access
```
Comment (gsmet) [inline docs/src/main/asciidoc/vault-auth.adoc:98]:
```suggestion
One drawback of this approach is that you expose a secret piece of information (i.e. the token) that can give access
```

File: docs/src/main/asciidoc/vault-auth.adoc [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+https://www.vaultproject.io/docs/concepts/response-wrapping[Response Wrapping] (which used to be known as
+_Cubbyhole Authentication_ in older versions of Vault). This principle is simple: instead of configuring the
+client token itself, we hide it inside a _Wrapping Token_, which we provide to the application. Upon startup,
+the application will unwrap the _Wrapping Token_, and fetch the real token from within. The additional level of
+security comes from the fact that the _Wrapping Token_ is short lived (from a few seconds to a few minutes;
+basically just enough to start and unwrap the token), and can be unwrapped *only once*.
+If the _Wrapping Token_ gets stolen and unwrapped, we will notice immediately because the legitimate application
+will get an error saying that the token is invalid.
+
+With that in mind, let’s create a new token and wrap it inside a _Wrapping Token_ with a ttl of 1 minute:
```
Comment (gsmet) [inline docs/src/main/asciidoc/vault-auth.adoc:113]:
```suggestion
With that in mind, let’s create a new token and wrap it inside a _Wrapping Token_ with a TTL of 1 minute:
```

File: docs/src/main/asciidoc/vault-auth.adoc [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+
+Now let’s use this wrapping token in the configuration:
+[source, properties, subs=attributes+]
+----
+quarkus.vault.authentication.client-token-wrapping-token={client-token-wrapping-token}
+----
+
+Compile and run the application *without the tests*, you should be able now to curl the private key `{private-key}` as before.
+
+Stop the application, and execute tests with `./mvnw test`. They should fail with error:
```
Comment (gsmet) [inline docs/src/main/asciidoc/vault-auth.adoc:135]:
```suggestion
Stop the application, and execute tests with `./mvnw test`. They should fail with the following error:
```

File: docs/src/main/asciidoc/vault-auth.adoc [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+----
+curl http://localhost:8080/hello/private-key
+----
+
+You should see: `{private-key}`.
+
+[NOTE]
+====
+Userpass supports response wrapping as well for the `password` property, although it is more
+unsual to use this approach as response wrapping typically involves a technical workflow,
```
Comment (gsmet) [inline docs/src/main/asciidoc/vault-auth.adoc:164]:
```suggestion
unusual to use this approach as response wrapping typically involves a technical workflow,
```
or usual? but I would say the former.

File: docs/src/main/asciidoc/vault-auth.adoc [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+quarkus.vault.authentication.app-role.secret-id-wrapping-token={secret-id-wrapping-token}
+----
+
+Finally, recompile the application without tests (otherwise you are going to consume the wrapping token),
+launch it and curl the `private-key` endpoint. As usual, you should see `{private-key}`.
+
+== Kubernetes Authentication
+
+Vault provides an integration with Kubernetes to allow containers to authenticate with Vault using their
+Kubernetes jwt token located in `/var/run/secrets/kubernetes.io/serviceaccount`.
```
Comment (gsmet) [inline docs/src/main/asciidoc/vault-auth.adoc:231]:
```suggestion
Kubernetes JWT token located in `/var/run/secrets/kubernetes.io/serviceaccount`.
```

File: docs/src/main/asciidoc/vault-auth.adoc [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+
+The setup is more involved than with the other auth methods because we need to allow Vault to talk
+to the master api to be able to validate the jwt token that the application will authenticate with.
```
Comment (gsmet) [inline docs/src/main/asciidoc/vault-auth.adoc:234]:
```suggestion
to the master API to be able to validate the JWT token that the application will authenticate with.
```

File: docs/src/main/asciidoc/vault-auth.adoc [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+
+=== auth-delegator
+
+The first step involves creating a `vault-auth-sa` service account with `auth-delegator`
+role that Vault will use to communicate with the master api.
```
Comment (gsmet) [inline docs/src/main/asciidoc/vault-auth.adoc:239]:
```suggestion
role that Vault will use to communicate with the master API.
```

File: docs/src/main/asciidoc/vault-auth.adoc [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+namespace: default
+roleRef:
+kind: ClusterRole
+name: system:auth-delegator
+apiGroup: rbac.authorization.k8s.io
+----
+
+And apply it: `kubectl apply -f vault-auth-k8s.yml`.
+
+Once the objects are created, we need to capture the jwt token of this service account,
```
Comment (gsmet) [inline docs/src/main/asciidoc/vault-auth.adoc:266]:
```suggestion
Once the objects are created, we need to capture the JWT token of this service account,
```

File: docs/src/main/asciidoc/vault-auth.adoc [hunk 13]
```diff
@@ -0,0 +1,433 @@
...
+    bound_service_account_names=vault-quickstart-sa \
+    bound_service_account_namespaces=default \
+    policies=vault-quickstart-policy \
+    ttl=1h
+----
+
+=== Deploy the application
+
+Add the following properties to the application (and remove any other authentication Vault property,
+plus replace `<host>` by the ip or name of the local host that is running the Vault and PostgresQL containers,
```
Comment (gsmet) [inline docs/src/main/asciidoc/vault-auth.adoc:318]:
```suggestion
plus replace `<host>` by the ip or name of the local host that is running the Vault and PostgreSQL containers,
```

**Review by gsmet — CHANGES_REQUESTED** [review]
Awesome! I added some very minor comments.

**Comment by gsmet** [comment]
@vsevel should I backport this document to 1.4? Are all the elements here in CR1? Do I need to backport more things? Thanks!

**Comment by vsevel** [comment]
dear all 

@lordofthejars @sberyozkin I modified the vault guide to refer to the vault auth guide in the following note: _Quarkus supports additional Vault Authentication methods. Check the {vault-auth-guide} for details._

I also added this precision in the `userpass` section of the vault auth guide: _Normally the userpass auth method should already be enabled from the {base-guide}. If not, execute the following commands now: ..._

> and also link to this doc from the main index page

are you talking about `index.adoc`? I thought this file was obsolete. Is there a place where I should list all vault guides?

@gsmet I have done all the suggested modifications. I have pushed and squashed so if everybody is fine, you can go ahead and merge. you can backport it too, since all the code is already in 1.4 (the last feature was _response wrapping_).

thanks for your comments.

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Thanks, great addition!

**Pull request merged**